woocommerce/google-listings-and-ads
# Add the partial auth feature of Google account to the reconnection subpage

### Changes proposed in this Pull Request:

Part of #1000

- Rename files in the *js/src/settings/reconnect-accounts* folder.
- Extract the `toScopeState` out as a util function and adjust some logic for checking the GLA required permission based on the current setup.
- Add the partial auth feature to the reconnection subpage

### Screenshots:

❗ Please note that these videos were recorded with local changes:
1. The same temporary hack as https://github.com/woocommerce/google-listings-and-ads/pull/1064#discussion_r736426466
2. Remove `hideIcon` props from all Google account card relevant components to show how it looks like after the #1065 is completed.

#### 🎥 . The original access check after reconnecting

💡  Even though accepts all permissions, asking user to disconnect is a priority action if user reconnects to an account that has no access to its previously connected GMC or Ads accounts.

https://user-images.githubusercontent.com/17420811/139220362-b018ddc3-3360-4f3e-b433-633a3db91566.mp4

#### 🎥 . Partial auth when granting without required permissions

https://user-images.githubusercontent.com/17420811/139225903-830687f9-8e5b-492a-b39b-8d44c7493a7d.mp4

#### 🎥 . If a user hasn't set up Google Ads account yet when reconnecting, then it could complete it without granting the ads permission
https://user-images.githubusercontent.com/17420811/139220392-a70276ff-8452-412c-b840-45a14b8fced7.mp4

### Detailed test instructions:

#### Test notes:

- 💡  The ads permission would be required if the Google Ads setup has been completed before entering reconnection process.
- 💡  Please ignore the problem of entering Google account choosing page first when going to Google OAuth page. It will be resolved once the dependency PR of #1063 in WCS is merged and deployed.

#### Pre-condition:

Use a Google account to complete the Setup MC flow. If the Setup MC flow has not been done, when you perform the following test step, the page will redirect you to the Setup MC page instead of reconnection page.

#### Test steps:

1. Disconnect Google account from the Connection Test page.
1. Go to a GLA page. It should redirect you to the reconnection page.
1. Connect with _another_ Google account with all/partial permissions, it should ask you to disconnect and reconnect due to the new connected account having no access to the previously linked Google Merchant Center account or Google Ads account. Click on the "Try another Google account" button to disconnect.
1. Connect with the original Google account without required permissions, it should re-ask authorization.
1. After re-authorizing the required permissions, it should redirect you to the dashboard page.

### Changelog entry

> Add - The partial auth feature of Google account to the reconnection subpage in the Settings page.
